targets/XPlatCppSdk/source/code/source/playfab/PlayFabAndroidHttpPlugin.cpp

```diff
@@ -0,0 +1,743 @@
+    {
+
+    }
+
+    PlayFabAndroidHttpPlugin::~PlayFabAndroidHttpPlugin()
+    {
+        threadRunning = false;
+
+        httpRequestMutex.lock();
+
+        if(workerThread)
+        {
+            httpRequestMutex.unlock();
+            workerThread->join();
+            httpRequestMutex.lock();
+        }
+        httpRequestMutex.unlock();
+    }
+
+    void PlayFabAndroidHttpPlugin::MakePostRequest(std::unique_ptr<CallRequestContainerBase> requestContainer)
+    {
+        std::shared_ptr<RequestTask> requestTask = nullptr;
+        try
+        {
+            requestTask = std::make_shared<RequestTask>();
+            requestTask->Initialize(requestContainer);
+        }
+        catch (const std::exception& ex)
+        {
+            PlayFabPluginManager::GetInstance().HandleException(ex);
+        }
+        catch (...)
+        {
+
+        }
+        if(requestTask != nullptr)
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            requestTask->state = RequestTask::Pending;
+            pendingRequests.push_back(std::move(requestTask));
+            if(workerThread == nullptr)
+            {
+                threadRunning = true;
+                workerThread = std::make_unique<std::thread>(&PlayFabAndroidHttpPlugin::WorkerThreadEntry, this);
+            }
+        } // UNLOCK httpRequestMutex
+    }
+
+    size_t PlayFabAndroidHttpPlugin::Update()
+    {
+        if (PlayFabSettings::threadedCallbacks)
+        {
+            throw std::runtime_error("You should not call Update() when PlayFabSettings::threadedCallbacks == true");
+        }
+
+        std::shared_ptr<RequestTask> requestTask = nullptr;
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            if (pendingResults.empty())
+            {
+                return pendingRequests.size();
+            }
+
+            requestTask = std::move(this->pendingResults[0]);
+            this->pendingResults.pop_front();
+        } // UNLOCK httpRequestMutex
+
+        HandleResults(*requestTask);
+
+        // activeRequestCount can be altered by HandleResults, so we have to re-lock and return an updated value
+        { // LOCK httpRequestMutex
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            return pendingRequests.size() + pendingRequests.size();
+        }
+    }
+
+    void PlayFabAndroidHttpPlugin::WorkerThreadEntry()
+    {
+        JavaVM* jvm;
+        if(!GetHelper().GetJavaVM(&jvm))
+        {
+            std::unique_lock<std::mutex> lock(httpRequestMutex);
+            threadRunning = false;
+            workerThread->detach();
+            workerThread = nullptr;
+            return ;
+        }
+        else
+        {
+            JNIEnv* jniEnv = nullptr;
+            jvm->AttachCurrentThread(&jniEnv, nullptr);
+
+            WorkerThread(static_cast<void*>(jniEnv));
+
+            jvm->DetachCurrentThread();
```

Comment:
The JVM does full garbage collection on the allocated JVM thread whenever DetatchCurrentThread is called.  Since we only appear to ever operate a single thread at a time, we might consider pooling the thread and keeping it around.  I have no data to back it up, but I strongly suspect that our pendingRequests will most of the time be empty, which means the vast majority of requests will incur the full cost of JVM thread allocation, construction, and garbage collection.

Reply:
this is depend on the perspective. I think the garbage collection isn't expensive than keep the resource. If this PlayFab is required time-sensitive response such as web server or game server, I will keep the thread alive. But I think PlayFab isn't time sensitive, and the keep resources until finishing isn't a good pattern if it can return resources to the system. But yes, some times to keep the resource is better, but in this case, if we keep the thread alive, we have to choose between the working thread should keep polling or suspend when there is no request. And the best case is suspending. but to make suspending implementation makes more complex. I think this pattern I implemented is good to keep simple and efficient.  and actually, the garbage collection isn't run every time, it is running when it is required. And, I checked the source code and it seems that to calling DetachCurrentThread() isn't related Java Object. This only related native resources. Anyway, such as using Kernel memory API can make kernel panic when the Kernel memory is completely exhausted. and on the Android easily reaches maximum memory usage.

---
In reply to: [275602778](https://github.com/PlayFab/SDKGenerator/pull/470#discussion_r275602778) [](ancestors = 275602778)

Reply:
Yes, it's certainly a tradeoff question.  I'd be interested to hear from @debhaldarMS on this point, since he would have a better insight as to typical usage than I would.  Still, if we can typically expect a relatively high frequency of API calls from a client, the memory usage may be required often enough to be considered permanent either way.  And if the frequency is at all high, the repetitive allocation and deallocation will only increase the risk of fragmentation, which may actually increase the chance that a larger allocation elsewhere will fail.